rust/parsing.rs

```diff
@@ -0,0 +1,945 @@
+                }
+                None => match &parsed.duration {
+                    Some(_) => {
+                        parsed.second_datetime = Some(datetime);
+                    }
+                    None => {
+                        parsed.datetime = Some(datetime);
+                    }
+                },
+            }
+
+            return Ok(());
+        }
+
+        datetime.has_date = true;
+        datetime.year = datetime.year * 100 + self.parse_integer(2, "year")?;
+
+        if self.current == '-' {
+            self.inc();
+            datetime.extended_date_format = true;
+
+            if self.current == 'W' {
+                // ISO week and day in extended format (i.e. Www-D)
+                self.inc();
+
+                let iso_week = self.parse_integer(2, "iso week")?;
+                let mut iso_day: u32 = 1;
+
+                if !self.end() && self.current != ' ' && self.current != 'T' {
+                    // Optional day
+                    if self.current != '-' {
+                        return Err(self.parse_error(format!(
+                            "Invalid character \"{}\" while parsing {}",
+                            self.current, "date separator"
+                        )));
+                    }
+
+                    self.inc();
+
+                    iso_day = self.parse_integer(1, "iso day")?;
+                }
+
+                match self.iso_to_ymd(datetime.year, iso_week, iso_day) {
+                    Ok((year, month, day)) => {
+                        datetime.year = year;
+                        datetime.month = month;
+                        datetime.day = day;
+                    }
+                    Err(error) => return Err(error),
+                }
+            } else {
+                /*
+                Month and day in extended format (MM-DD) or ordinal date (DDD)
+                We'll assume first that the next part is a month and adjust if not.
+                */
+                datetime.month = self.parse_integer(2, "month")?;
+
+                if !self.end() && self.current != ' ' && self.current != 'T' {
+                    if self.current == '-' {
+                        // Optional day
+                        self.inc();
+                        datetime.day = self.parse_integer(2, "day")?;
+                    } else {
+                        // Ordinal day
+                        let ordinal_day =
+                            (datetime.month * 10 + self.parse_integer(1, "ordinal day")?) as i32;
+
+                        match self.ordinal_to_ymd(datetime.year, ordinal_day, false) {
+                            Ok((year, month, day)) => {
+                                datetime.year = year;
+                                datetime.month = month;
+                                datetime.day = day;
+                            }
+                            Err(error) => return Err(error),
+                        }
```

Comment:
this can be simplified using the `?` operator
```suggestion
                        let (year, month, day) =
                            self.ordinal_to_ymd(datetime.year, ordinal_day, true)?;

                        datetime.year = year;
                        datetime.month = month;
                        datetime.day = day;
```